### ISSUE TITLE
Foreign keys are not supported while evaluating the tool

### ISSUE THREAD

**Issue by avifatal** [skeema/skeema#18]
Hi
Im evaluating the tool.
I have run it with debug. nothing working.

this is the output:

```
2018-03-05 22:17:39 [INFO]  Generating diff of 127.0.0.1:3306 icoserver vs /var/www/schemas/icoserver/*.sql
2018-03-05 22:17:46 [WARN]  Skipping table customers: unable to generate ALTER TABLE due to use of unsupported features
2018-03-05 22:17:46 [DEBUG] --- Skeema-expected
2018-03-05 22:17:46 [DEBUG] +++ MySQL-actual
2018-03-05 22:17:46 [DEBUG] @@ -33 +33,6 @@
2018-03-05 22:17:46 [DEBUG] -  KEY `customers_primary_tracking_id_foreign` (`primary_tracking_id`)
2018-03-05 22:17:46 [DEBUG] +  KEY `customers_primary_tracking_id_foreign` (`primary_tracking_id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `customers_country_id_foreign` FOREIGN KEY (`country_id`) REFERENCES `countries` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `customers_customer_user_id_foreign` FOREIGN KEY (`customer_user_id`) REFERENCES `users` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `customers_ip_country_id_foreign` FOREIGN KEY (`ip_country_id`) REFERENCES `countries` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `customers_last_status_by_user_id_foreign` FOREIGN KEY (`last_status_by_user_id`) REFERENCES `users` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `customers_primary_tracking_id_foreign` FOREIGN KEY (`primary_tracking_id`) REFERENCES `trackings` (`id`)
2018-03-05 22:17:46 [WARN]  Skipping table referrals: unable to generate ALTER TABLE due to use of unsupported features
2018-03-05 22:17:46 [DEBUG] --- Skeema-expected
2018-03-05 22:17:46 [DEBUG] +++ MySQL-actual
2018-03-05 22:17:46 [DEBUG] @@ -13 +13,2 @@
2018-03-05 22:17:46 [DEBUG] -  KEY `referrals_traffic_source_id_foreign` (`traffic_source_id`)
2018-03-05 22:17:46 [DEBUG] +  KEY `referrals_traffic_source_id_foreign` (`traffic_source_id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `referrals_traffic_source_id_foreign` FOREIGN KEY (`traffic_source_id`) REFERENCES `traffic_sources` (`id`)
2018-03-05 22:17:46 [WARN]  Skipping table trackings: unable to generate ALTER TABLE due to use of unsupported features
2018-03-05 22:17:46 [DEBUG] --- Skeema-expected
2018-03-05 22:17:46 [DEBUG] +++ MySQL-actual
2018-03-05 22:17:46 [DEBUG] @@ -24 +24,5 @@
2018-03-05 22:17:46 [DEBUG] -  KEY `trackings_referral_id_foreign` (`referral_id`)
2018-03-05 22:17:46 [DEBUG] +  KEY `trackings_referral_id_foreign` (`referral_id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `trackings_customer_id_foreign` FOREIGN KEY (`customer_id`) REFERENCES `customers` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `trackings_ip_country_id_foreign` FOREIGN KEY (`ip_country_id`) REFERENCES `countries` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `trackings_referral_id_foreign` FOREIGN KEY (`referral_id`) REFERENCES `referrals` (`id`),
2018-03-05 22:17:46 [DEBUG] +  CONSTRAINT `trackings_traffic_source_id_foreign` FOREIGN KEY (`traffic_source_id`) REFERENCES `traffic_sources` (`id`)
2018-03-05 22:17:46 [INFO]  127.0.0.1:3306 icoserver: diff complete
```

I dont think I have special case here, nothing besides tables with foreign keys.
I can't really understand the drbug data.
Thanks

**Comment by evanelias** [comment]
Thanks for trying out Skeema. Foreign keys are not supported yet. This is noted in the [docs](https://github.com/skeema/skeema/blob/master/doc/requirements.md#unsupported-for-alters). 

Support for foreign keys has been started via some great work by a contributor (see #8 and skeema/tengo#1), but is not yet merged into master due to hitting a rather complex blocker (see skeema/tengo#2). Also see #11 for background on why foreign key usage in MySQL is generally discouraged.

Please note that Skeema is designed for use with external OSC tools, which either [have major caveats with foreign keys](https://www.percona.com/doc/percona-toolkit/LATEST/pt-online-schema-change.html#cmdoption-pt-online-schema-change-alter-foreign-keys-method) or simply [don't support them at all](https://github.com/github/gh-ost/blob/master/doc/requirements-and-limitations.md#limitations).

In summary, fully supporting foreign keys would be quite complex, and large MySQL shops don't use them anyway. Altogether, that means it is hard for me to justify dedicating the development time on my end, at least in a high-priority fashion. Please understand that Skeema is a free open-source tool without corporate affiliation or backing.

**Comment by avifatal** [comment]
Thanks for the detailed response.
So, if I give up on foreign keys, most of the features will be supported?

I know that foreign keys are no go in big databases...

Thanks

**Comment by evanelias** [comment]
Hopefully yes, but it depends on what your tables are doing. Please see https://github.com/skeema/skeema/blob/master/doc/requirements.md#unsupported-features for a list of unsupported features.
